contributors/design-proposals/key-set-as-merge-key.md

```diff
@@ -0,0 +1,338 @@
+# Key Set as Merge Key in Strategic Merge Patch
+
+## Abstract
+
+Support a key set as the merge key in Strategic Merge Patch.
+
+## Motivation
+
+*Background*: In Strategic Merge Patch, Merge Key is the key to distinguish the entries in the list of non-primitive types.
+It must always be present and unique to perform the merge on the list of non-primitive types,
+and will be preserved.
+
+The current implementation requires a single field that uniquely identifies each element in a list.
+For some element Kinds, the identity is defined using multiple fields.
+An [example](https://github.com/kubernetes/kubernetes/issues/39188 of this is the service.spec.ports, which is identified by `protocol` and `port`.
+We need to support a key set as Merge Key. A key set must contain at least 1 key.
+
+## Proposed Change
+
+### API Change
+
+We will update the struct tags for the API resources whose Merge Key is not guaranteed to be always
+present and unique. The keys will be seperated by ",", i.e. `patchMergeKey:"<key1>,<key2>,<key3>"`.
+
+E.g. `Ports` in `ServiceSpec`.
+```go
+type ServiceSpec struct {
+  // Change patchMergeKey from "port" to "name|port"
+  Ports []ServicePort `json:"ports,omitempty" patchStrategy:"merge" patchMergeKey:"name,port" protobuf:"bytes,1,rep,name=ports"`
+  ...
+}
+```
+
+All the impacted APIs are listed in section [Impacted APIs](#impacted-apis)
+
+### Open API
+
+Update Open API schema to have a string that contains multiple keys for `patchMergeKey`.
+The change should be trivial.
+
+### Strategic Merge Patch pkg
+
+Entries are considered as the same if and only if all the keys in the key set are identical.
+We allow keys to be missing in the key set as long as it is not empty.
+
+Take `Ports` as an example:
+
+Suppose we are using `name` and `port` as merge key as mentioned in section [API Change](#api-change)
+
+We have live config:
+```yaml
+spec:
+  type: NodePort
+  ports:
+    - protocol: UDP
+      # name is missing here
+      port: 30420
+      nodePort: 30420
+```
+
+We want to add another port, and the user's local config file becomes:
+```yaml
+spec:
+  type: NodePort
+  ports:
+    - protocol: UDP
+      port: 30420
+      name: udpport
+      nodePort: 30420
+    - protocol: TCP
+      port: 30420
+      name: tcpport
+      nodePort: 30420
+```
+
+The patch we send should be:
+```yaml
+spec:
+  type: NodePort
+  ports:
+    # the entry with key set {port: 30420} is considered missing in local config file
+    - port: 30420
+      $patch: delete
+    # the entry with key set {name: udpport, port: 30420} is considered as a new entry
+    - protocol: UDP
+      port: 30420
+      name: udpport
+      nodePort: 30420
+    # the entry with key set {name: tcpport, port: 30420} is a new entry
+    - protocol: TCP
+      port: 30420
+      name: tcpport
+      nodePort: 30420
+```
+
+### Docs
+
+Document what the developer should consider when adding an API with `mergeKey`.
+
+## Version Skew
```

Comment:
@thockin FYI

To clarify, this is an issue generally with changing merge keys, not specifically with supporting tuples as merge keys.

A couple thoughts:
- The ship may have sailed on this already, but I would think that the client should sent the merge keys and merge strategies as part of the patch request so it could tell the server how the patch request is intended to be merged.  
- It maybe possible to introduce additional keys for fields while keeping backwards compatibility.  Would need to think through the combinatorics, but I am wondering if some of the merge key tuple is missing from the patch request (e.g. the new part), if the server could fallback on the remaining pieces as the merge key.
- If merge keys really are "forever", and we needed to change the one referenced in this case, would this then require bumping the version for everything that has a PodTemplate in it - e.g. every controller type?  The cure may be worse than the disease in this case.
- Given that we don't publish what the merge keys actually are for resource kinds in any user facing documentation besides the `types.go` files (correct me if I am wrong here), are they considered first class pieces of the API, or are they an implementation detail?  Does our API definition include metadata about fields that is only defined in code?  Do we have any guess as to which clients besides kubectl calculate patch requests?